Symint-ify _upsample_bilinear2d_aa

gmagogsfm opened the pull request:
Differential Revision: D42935541

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -2639,6 +2639,21 @@ def meta_bucketize(self, boundaries, *, out_int32=False, right=False):
+    full_output_size = upsample_common_check(
+        input.size(), output_size, num_spatial_dims=2
+    )
+    assert input.numel() != 0 or all(
```
use `check()` instead of assert

gmagogsfm replied:
Done.

ezyang commented on c10/core/SymFloat.cpp:
```diff
@@ -88,4 +90,12 @@ double SymFloat::guard_float(const char* file, int64_t line) const {
 
+SymInt SymFloat::floor() const {
+  if (!is_symbolic()) {
+    return int(data_);
```
use floor, not int; when negative floor goes to negative infinity

gmagogsfm replied:
Changed, forgot that they represent more than just sizes

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -2639,6 +2639,21 @@ def meta_bucketize(self, boundaries, *, out_int32=False, right=False):
     ).contiguous()
 
 
+@register_meta(aten._upsample_bilinear2d_aa.default)
```
So what's going on, are you doing this in Python or in C++ lol. You don't need both.

gmagogsfm replied:
I think the logic in UpSampleBilinear2d.cpp is not a Meta kernel, instead it looks more like a C++ binded to Python API. Line 180 of UpSampleBilinera2d.cpp invokes this Python Meta kernel. PTAL.

ezyang replied:
I see, there's a vec and non-vec overload, ok

ezyang reviewed (changes requested):
see comments. I'm also expecting a test to start passing

gmagogsfm reviewed (commented):
@ezyang  PTAL

gmagogsfm commented:
> see comments. I'm also expecting a test to start passing

which test? do you mean a new test case should be added to cover the meta kernel?

ezyang commented on aten/src/ATen/native/UpSample.cpp:
```diff
@@ -30,6 +30,29 @@ TORCH_API c10::SmallVector<int64_t, 3> compute_output_size(
+    }
+    return ret;
+  }
+  TORCH_CHECK(false, "Must specify exactly one of output_size and scale_factors");
```
We have an implementation of this in Python at torch/_decomp/decompositions.py:def upsample_compute_output_size(input_size, output_size, scale_factors) for the upsample decomps, so it feels like it would be nicer if the aa composite was ported to Python, like the other decomps are, then you wouldn't need to dupe this function.

ezyang commented on aten/src/ATen/native/UpSample.cpp:
```diff
@@ -30,6 +30,29 @@ TORCH_API c10::SmallVector<int64_t, 3> compute_output_size(
+    c10::SmallVector<SymInt, 3> ret;
+    for (const auto i : c10::irange(spatial_dimensions)) {
+      const auto scaled_float = SymFloat(input_size[i+2]) * scale_factors.value()[i];
+      ret.push_back(scaled_float.floor());
```
Ah, but in the old code, this is not a floor, this is an int cast. We have a separate `sym_int` for expressly this purpose! Fortunately in this case ret is guaranteed to be a size and positive so it is a moot point,

ezyang reviewed (approved):
OK, but don't forget to fix lint

gmagogsfm commented:
> > see comments. I'm also expecting a test to start passing
> 
> which test? do you mean a new test case should be added to cover the meta kernel?

@ezyang , could you respond here? what test?

ezyang commented:
When a kernel gets symintified, typically the symbolic exhaustive test in test_proxy_tensor.py starts xpassing. If it is not please investigate

gmagogsfm committed:
Syminti-fy _upsample_bilinear2d_aa (#94516)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/94516

Differential Revision: D42935541

fbshipit-source-id: 5e00af9b06b558f610cde77d9c6e3bbd77f506fb

gmagogsfm commented:
Abandoning this approach to use Python decomp
